src/zkp/piaffg.rs

```diff
@@ -604,12 +608,437 @@ mod tests {
+                .unwrap();
+            bad_proof.random_add_coeff_ciphertext_prover = ciphertext;
+            assert_ne!(
+                bad_proof.random_add_coeff_ciphertext_prover,
+                proof.random_add_coeff_ciphertext_prover
+            );
+            assert!(bad_proof.verify(input, &(), &mut transcript()).is_err());
+
+            // Swap random_mult_coeff_exp with a random [`CurvePoint`]
+            let mut bad_proof = proof.clone();
+            let mask = random_plusminus_by_size(&mut rng, ELL);
+            bad_proof.random_mult_coeff_exp = CurvePoint::GENERATOR.multiply_by_scalar(&mask)?;
+            assert_ne!(bad_proof.random_mult_coeff_exp, proof.random_mult_coeff_exp);
+            assert!(bad_proof.verify(input, &(), &mut transcript()).is_err());
+
+            // Swap masked_add_coeff_nonce_prover with a random [`MaskedNonce`]
+            let mut bad_proof = proof.clone();
+            bad_proof.masked_add_coeff_nonce_prover =
+                MaskedNonce::random(&mut rng, input.prover_encryption_key.modulus());
+            assert_ne!(
+                bad_proof.masked_add_coeff_nonce_prover,
+                proof.masked_add_coeff_nonce_prover
+            );
+            assert!(bad_proof.verify(input, &(), &mut transcript()).is_err());
+
+            // Swap masked_add_coeff_nonce_prover with a random [`MaskedNonce`]
+            let mut bad_proof = proof.clone();
+            bad_proof.masked_add_coeff_nonce_prover =
+                MaskedNonce::random(&mut rng, input.prover_encryption_key.modulus());
+            assert_ne!(
+                bad_proof.masked_add_coeff_nonce_prover,
+                proof.masked_add_coeff_nonce_prover
+            );
+            assert!(bad_proof.verify(input, &(), &mut transcript()).is_err());
+
+            // Swap masked_mult_coeff_commit_randomness with a random [`MaskedRandomness`]
+            let mut bad_proof = proof.clone();
+            bad_proof.masked_mult_coeff_commit_randomness = bad_randomness.as_masked().to_owned();
+            assert_ne!(
+                bad_proof.masked_mult_coeff_commit_randomness,
+                proof.masked_mult_coeff_commit_randomness
+            );
+            assert!(bad_proof.verify(input, &(), &mut transcript()).is_err());
+
+            // Swap masked_add_coeff_commit_randomness with a random [`MaskedRandomness`]
+            let mut bad_proof = proof.clone();
+            bad_proof.masked_add_coeff_commit_randomness = bad_randomness.as_masked().to_owned();
+            assert_ne!(
+                bad_proof.masked_add_coeff_commit_randomness,
+                proof.masked_add_coeff_commit_randomness
+            );
+            assert!(bad_proof.verify(input, &(), &mut transcript()).is_err());
+
+            Ok(())
+        };
+        with_random_paillier_affg_proof(rng2, &x, &y, f)?;
+        Ok(())
+    }
+
     #[test]
     fn test_paillier_affg_proof() -> Result<()> {
```

Comment:
I actually see this proof does some range testing, too. I prefer having a separate named proof that tests the range stuff like you have but just noting the duplication here.